File: lex.go
package main

import (
	"fmt"
	"log"
)

type tokenType int

const (
	// Single Charicter Tokens
	tokenOpenParen tokenType = iota
	tokenCloseParen
	tokenOpenBrace
	tokenCloseBrace
	tokenSemicolon

	// Literals
	tokenIdentfier
	tokenConstant

	// Keywords
	tokenInt
	tokenVoid
	tokenReturn

	// misc
	tokenEOF
)

var keywords = map[string]tokenType{
	"int":    tokenInt,
	"void":   tokenVoid,
	"return": tokenReturn,
}

type token struct {
	tokenType tokenType
	literal   any
	line      int
}

type Lexer struct {
	tokens  []token
	source  string
	start   int
	current int
	line    int
}

func NewLexer(source string) *Lexer {
	l := &Lexer{
		source: source,
	}
	return l
}

func (l *Lexer) isAtEnd() bool {
	return l.current >= len(l.source)
}

func (l *Lexer) lex() error {
	for {
		if l.isAtEnd() {
			return nil
		}

		l.start = l.current
		l.scanToken()
	}
}

func (l *Lexer) scanToken() error {
	c := l.advance()
	switch c {
	case '(':
		l.addToken(tokenOpenParen, "(")
	case ')':
		l.addToken(tokenCloseParen, ")")
	case '{':
		l.addToken(tokenOpenBrace, "{")
	case '}':
		l.addToken(tokenCloseBrace, "}")
	case ';':
		l.addToken(tokenSemicolon, ";")

	// Ignore whitespace
	case ' ', '\t', '\r':
		return nil

	case '\n':
		l.line += 1
	default:
		if isDecimal(c) {
			l.number()
		} else if isAlpha(c) {
			l.identifier()
		} else {
			return fmt.Errorf("invalid char: %s", string(c))
		}
	}
	return nil
}

func (l *Lexer) number() {
	for {
		if isDecimal(l.peek()) {
			l.advance()
		} else {
			break
		}
	}

	if !l.isAtEnd() && isAlpha(l.peek()) {
		log.Fatalf("Invalid number at line %d", l.line)
	}
	l.addToken(tokenConstant, l.source[l.start:l.current])
}

func (l *Lexer) identifier() {
	for {
		if isAlphaNumeric(l.peek()) {
			l.advance()
		} else {
			break
		}
	}

	text := l.source[l.start:l.current]
	keyword, ok := keywords[text]
	if !ok {
		l.addToken(tokenConstant, l.source[l.start:l.current])
	} else {
		l.addToken(keyword, l.source[l.start:l.current])
	}
}

func (l *Lexer) addToken(tokenType tokenType, literal any) {
	l.tokens = append(l.tokens, token{tokenType: tokenType, literal: literal, line: l.line})
}

func (l *Lexer) advance() byte {
	// TODO: check if is at end like peek()
	c := l.source[l.current]
	l.current += 1
	return c
}

func (l *Lexer) peek() byte {
	// TODO: find some other way to indicate end of file
	//if l.isAtEnd() {
	//  return '\0'
	//}
	return l.source[l.current]
}

func isDecimal(c byte) bool {
	return '0' <= c && c <= '9'
}

func isAlpha(c byte) bool {
	return (c >= 'a' && c <= 'z') ||
		(c >= 'A' && c <= 'Z') ||
		c == '_'
}

func isAlphaNumeric(c byte) bool {
	return isAlpha(c) || isDecimal(c)
}

File: lex.py
from enum import Enum, auto
from os import curdir
from typing import Any
class TokenType(Enum):
    # Single-character tokens
    OPEN_PAREN = auto()
    CLOSE_PAREN = auto()
    OPEN_BRACE = auto()
    CLOSE_BRACE = auto()
    SEMICOLON = auto()

    # Literals
    IDENTIFIER = auto()
    CONSTANT = auto()

    # Keywords
    INT = auto()
    VOID = auto()
    RETURN = auto()

keywords = {"int": TokenType.INT, "void": TokenType.VOID, "return": TokenType.RETURN}

class Token:
    def __init__(self, type: TokenType, lexeme: str, literal: Any, line: int) -> None:
        self.type = type
        self.lexeme = lexeme
        self.literal = literal
        self.line = line

    def __str__(self) -> str:
        return f"{self.type.name} {self.lexeme} {self.literal}"

class Lexer:
    def __init__(self, source: str) -> None:
        self.source: str = source
        self.tokens = []

        self.start = 0
        self.current = 0
        self.line = 1

    def isAtEnd(self):
        return self.current >= len(self.source)

    def scanTokens(self):
        while not self.isAtEnd():
            self.start = self.current
            self.scanToken()
            
        return self.tokens
    
    def scanToken(self):
        c = self.advance()
        match c:
            case "(":
                self.add_token(TokenType.OPEN_PAREN, None)
            case ")":
                self.add_token(TokenType.CLOSE_PAREN, None)
            case "{":
                self.add_token(TokenType.OPEN_BRACE, None)
            case "}":
                self.add_token(TokenType.CLOSE_BRACE, None)
            case ";":
                self.add_token(TokenType.SEMICOLON, None)

            # Ignore whitespace
            case " ":
                pass
            case "\t":
                pass
            case "\r":
                pass

            case "\n":
                self.line += 1

            case _:
                if c.isdigit():
                    self.number()
                elif c.isalnum():
                    self.identifier()
                else:
                    raise Exception("Invalid char")

    def advance(self):
        c = self.source[self.current]
        self.current += 1
        return c

    def peek(self) -> str:
        if self.isAtEnd(): 
            return '\0'
        return self.source[self.current]

    def add_token(self, type: TokenType, literal: Any):
        text = self.source[self.start:self.current]
        self.tokens.append(Token(type, text, literal, self.line))
    
    def number(self):
        while self.peek().isdigit():
            self.advance()
        
        if not self.isAtEnd() and self.peek().isalpha():
            raise Exception(f"Invalid numeric format at line {self.line}: unexpected '{self.peek()}' after number")
    
        self.add_token(TokenType.CONSTANT, int(self.source[self.start:self.current]))
    
    def identifier(self):
        while self.peek().isalnum():
            self.advance()

        text = self.source[self.start:self.current]
        type = keywords[text] if text in keywords.keys() else TokenType.IDENTIFIER
        self.add_token(type, None)

